nwbinspector/nwbinspector.py

```diff
@@ -34,6 +35,54 @@ def default(self, o):
             return super().default(o)
 
 
+def organize_messages_by_file(messages):
+    files = sorted(set(message.file for message in messages))
+    out = {}
+    for file in files:
+        out[file] = {}
+        file_messages = list(filter(lambda x: x.file == file, messages))
+        for importance in sorted(set(message.importance for message in file_messages), key=lambda x: -x.value):
+            out[file][importance] = sorted(
+                filter(lambda x: x.importance == importance, file_messages),
+                key=lambda x: -x.severity.value,
+            )
+    return out
```

Comment:
After testing this, one thing to note is the time complexity involved in applying a `filter` that iterates a list checking for a potentially small number of boolean matches is sub-optimal. Here's a self-contained comparison that can be applied to any directory with a decent number of files in it

```python
"""Simple script for testing performance of two iteration methods for NWBInspector."""
from time import time
from pathlib import Path

import numpy as np

files = [x.name for x in Path(".").iterdir()]
nfiles = len(files)
messages = [
    dict(msg="foo", file=files[np.random.randint(low=0, high=nfiles)])
    for _ in range(nfiles * 50)
]

start = time()
out1 = {file: list(filter(lambda x: x["file"] == file, messages)) for file in files}
end = time()
time1 = end - start

start = time()
out2 = {file: list() for file in files}
for message in messages:
    out2[message["file"]] = message
end = time()
time2 = end - start
```

For my system at least, the binning method (proposed option #2) is 10x faster.

Reply:
As such, I was going to refactor this type of functionality like

```python
def organize_messages_by_file(messages: List[InspectorMessage]):
    """Order InspectorMessages by file name then importance."""
    files = natsorted(set(message.file for message in messages))
    messages_by_file = {file: list for file in files}
    out = copy(messages_by_file)
    for message in messages:
        messages_by_file[message.file] = message
    for file, messages_for_file in messages_by_file.items():
        realized_importance = sorted(set([message.importance for message in messages_for_file]), key=lambda x: -x.value)
        messages_per_importance = {importance: [] for importance in realized_importance}
        for message in messages_for_file:
            messages_per_importance[message.importance]
        out[message.file] = sorted(messages_per_importance, key=lambda x: -x.severity.value)
    return out
```

Basically, the more pre-sorting you can apply to the output structure ahead of actual iteration, the better.

@bendichter What do you think? Any other suggestions for improvement?

Reply:
@CodyCBakerPhD how about this as a general solution

```python
def hier_org(objects, levels=["attr1", "attr2", "attr3"]):
    unique_attrs = sorted(set(getattr(object, levels[0]) for object in objects))
    if len(levels) > 1:
        return {attr: hier_org([obj for obj in objects if getattr(obj, levels[0]) == attr], levels[1:])
            for attr in unique_attrs}
    else:
        return {attr: [obj for obj in objects if getattr(obj, levels[0]) == attr] for attr in unique_attrs}
    

# test
from collections import namedtuple

Obj = namedtuple("Obj", "a b c")

objects = []
for a in range(10):
    for b in range(10):
        for c in range(10):
            objects.append(Obj(a=a,b=b,c=c))
            
hier_org(objects, ["a", "b", "c"])
```

Reply:
That looks like it should cover my next question if implemented correctly, which was to have a single `organize_messages` function that had an argument that could dynamically specify the order in which we want things organized (`["file", "importance"]`, or `["importance", "neurodata_type", "importance"]`, etc.)

I'll play around with this a bit and get back to you!

Reply:
@bendichter That code had some great notions of recurrence logic in it, but it still suffered from the same nested iteration as before due to the line

```python
        return {attr: hier_org([obj for obj in objects if getattr(obj, levels[0]) == attr], levels[1:])
            for attr in unique_attrs}
```

which still iterations over **every** `obj` for **each** `unique_aatr` to extract the subset that correspond to a particular `attr`. A time complexity analysis of this shows a worst-case time of 

![image](https://user-images.githubusercontent.com/51133164/157324261-069da4e6-b7bf-4c25-9346-07abc105c8f0.png)

which will tend to be the two largest values of 

![image](https://user-images.githubusercontent.com/51133164/157324646-28271d7a-a116-4578-9390-4981bf72d645.png)

![image](https://user-images.githubusercontent.com/51133164/157325674-337829d3-c025-4944-87d2-70ddb972cb09.png)

I took this logic and made two versions seen on this commit on another branch: https://github.com/NeurodataWithoutBorders/nwbinspector/commit/21a30aadaf1e0f51cc5bfd183b79fa181315a233

First version 'fancy' is pretty much that same exact code but built on our objects +/- some Enum logic for sorting.

Second version 'efficient' is my binned approach that separates the dictionary construction from the data filtering, though it needs some helper functions for indexing arbitrarily nested dictionaries. This decoupling approach improves the time complexity to

![image](https://user-images.githubusercontent.com/51133164/157325281-f9e4592a-57b8-4ea3-af99-8aaf70584b0f.png)

To test performance, I downloaded DANDISet `000004` and set up the following

```python
messages = list()
for folder in Path(".../000004").iterdir():
    messages.extend(inspect_all(path=folder))
```

which gives us a list of 611 InspectorMessages. Now

```python
levels = ["file", "importance"]
 # just for testing, technically any order and combination of attributes for an InspectorMessage can be used

start = time()
res1 = fancy_organize_messages(messages=messages, levels=levels)
end = time()
time1 = end - start

start = time()
res2 = efficient_organize_messages(messages=messages, levels=levels)
end = time()
time2 = end - start
```

As before, the 'efficient' method is clocking much faster (about 6x on the case of real data), so I'll proceed with adding tests and integrating that method.

Overall very cool advancement that allows arbitrary user choice of output structure, if they even want any.

Reply:
how about this:

```python
from collections import defaultdict

def hier_org(objects, levels, reverse=None):
    if reverse is None:
        reverse = [False] * len(levels)
    unique_attrs = sorted(set(getattr(object, levels[0]) for object in objects))
    if reverse[0]:
        unique_attrs = unique_attrs[::-1]
    objects_by_attr = defaultdict(list)
    [objects_by_attr[getattr(obj, levels[0])].append(obj) for obj in objects]
    if len(levels) > 1:
        return {attr: hier_org(objects_by_attr[attr], levels[1:], reverse[1:]) for attr in unique_attrs}
    else:
        return {attr: objects_by_attr[attr][0] for attr in unique_attrs}
    
    

# test
from collections import namedtuple

Obj = namedtuple("Obj", "a b c")

objects = []
for a in range(3):
    for b in range(3):
        for c in range(3):
            objects.append(Obj(a=a, b=b, c=c))
            
hier_org(objects, ["a", "b", "c"], [True, False, True])
```